### ISSUE TITLE
Choice of y axis intercept

### ISSUE THREAD

**Issue by e-pet** [meeliskull/prg#10]
Hi everyone,

so this is apparently not being maintained anymore (?); I'm just leaving this here as documentation, also in light of recent discussions to add PRG analysis to other packages [[1]](https://github.com/scikit-learn/scikit-learn/pull/24121), [[2]](https://github.com/Lightning-AI/torchmetrics/issues/1775).

Consider the following case with a single positive example that does _not_ have the highest score associated with it:
```python
target = np.array([0, 0, 0, 0, 0, 0, 0, 0, 0, 1])
y_score = np.array([0.01, 0.02, 0.03, 0.04, 0.05, 0.06, 0.07, 0.08, 0.09, 0.05])
```

This yields
```python
prec, rec, thresh = sklearn.metrics.precision_recall_curve(target, y_score, drop_intermediate=True)
prec  # array([0.1       , 0.16666667, 0.        , 0.        , 1.        ])
rec  # array([1., 1., 0., 0., 0.])
thresh  # array([0.01, 0.05, 0.06, 0.09])
```
as would be expected: 
- We can use threshold = 0.05 and get recall = 1 and precision = 0.166... (Note: better than the always-positive classifier)
- Any threshold larger than that gives recall = 0 and precision = 0.
- [By default](https://scikit-learn.org/dev/modules/generated/sklearn.metrics.precision_recall_curve.html), an artificial first point at precision = 1 and recall = 0 is added. (I'm not really convinced this makes any sense, but that is a separate discussion.)

![image](https://github.com/user-attachments/assets/7a5e325e-1a8f-4cbf-981e-92919ecb5d75)

So: this model essentially has a single meaningful PR point at recall = 1 and precision = 0.166..., beating the always-positive baseline (which has recall = 1 and precision = 0.1). Arguably, there is a second 'real' datapoint at (0, 0).

If I now take the formulae from the PRG paper, I get
```python
baseline = target.sum() / len(target)
precG = (prec - baseline) / (1 - baseline) / prec  # array([0.        , 0.44444444,       -inf,       -inf, 1.        ])
recG = (rec - baseline) / (1 - baseline) / rec  # array([  1.,   1., -inf, -inf, -inf])
```
where the single meaningful data point is translated into (precG = 0.444, recG = 1). So far, so good.

However, if I now calculate the AUPRG, I obtain
```python
auprg = calc_auprg(create_prg_curve(target, y_score))  # -1.555555555555555
```
suggesting that this model is *worse* than the always-positive baseline.

I'm not entirely sure whether this is technically _wrong_, but at least it seems undesirable to me?

If I look at the computed precG, recG values, I see that there are various 'artificial' datapoints:
```python
prg_curve['recall_gain']  # array([-inf, -inf, -inf, -inf, -inf, 0., 0.88888889, 1., 1., 1., 1., 1.])
prg_curve['precision_gain']  # array([-inf, -inf, -inf, -inf, nan, -3.55555556, 0., 0.44444444, 0.33333333, 0.22222222, 0.11111111, 0.])
prg_curve['is_crossing']  # array([0., 0., 0., 0., 0., 1., 1., 0., 0., 0., 0., 0.])
```
where `is_crossing` [apparently indicates](https://github.com/meeliskull/prg/blob/73ce510cb1b9e4675de831d82d9e43d756a66218/Python_package/prg/prg.py#L212) artificial crossing points that are always added ["because calc_auprg otherwise gives the wrong result"](https://github.com/meeliskull/prg/blob/73ce510cb1b9e4675de831d82d9e43d756a66218/Python_package/prg/prg.py#L190C50-L190C101). I could not find more details about these crossing points anywhere.

Given that, as stated above, there is basically a single meaningful datapoint here at recG=1, the question boils down to the choice of the correct y axis intercept, I believe: the whole remaining curve (and thus AUPRG) is then obtained from linear interpolation between these two points.

The y axis intercept ("y0" in the paper) is the point where recG = 0; however, this point is simply undefined for the given classifier (there is no point where rec = baseline, and it also cannot be obtained from interpolation in gain space since there is no well-defined point further left - for rec=0, prec=0 the gains cannot be calculated) and I'm not sure there is a way to obtain it in any meaningful way.

I'm not yet entirely sure what to learn from this:
- The choice of the y axis intercept is crucial; I could not find any documentation on how this is done in the package and have not yet looked deeply enough into the implementation
- It seems that results are questionable / unreliable in specific circumstances like the above, but I'm not yet entirely sure how to characterize these situations (e.g. in order to produce a warning). One check could be whether there are 'real' datapoints with recall <= baseline, but this would flag _a lot_ of practical use cases, I assume - especially, since PRG analysis is often recommended particularly for the case of strong class imbalance.

**Comment by e-pet** [comment]
Some follow-up thoughts.

To obtain a finite precG, we need prec > 0. This is the case if there is at least one positive sample with score >= the chosen threshold. At this point, we have recall = num_pos_samples_with_this_score / num_pos_samples. *This gives us the leftmost 'real' point in PRG space; anything to the left of this is simply undefined and no area under the PRG curve can be meaningfully calculated there.*

Now consider a relatively typical case with proportion of positives = 0.01 and num_pos_samples = 50. In that case, the leftmost 'real' point in PRG space is *at best* (in the case where num_pos_samples_with_this_score = 1) at recG = 0.51! For a more extreme case with num_pos_samples = 5, we even get recG = 0.96.

At least to my current understanding, the only real solution would be to compute a *partial* AUPRG with a pretty high recG cutoff. But I guess then the question becomes: how interpretable and meaningful is such a measure, really?

**Comment by e-pet** [comment]
If anyone should ever stumble upon this: I ended up implementing this (basically from scratch) in https://github.com/FraunhoferMEVIS/meval/tree/main, specifically here (includes a lot of comments + explanations): https://github.com/FraunhoferMEVIS/meval/blob/6b166e9ec5b893a6f3b29fb82ffb679fe9d0b0fc/meval/metrics/_metrics.py#L290

Also cf. the accompanying [paper](https://arxiv.org/abs/2512.17409) in which we also discuss some of the subtleties.
